emission/storage/decorations/trip_queries.py

```diff
@@ -73,66 +76,87 @@ def get_stops_for_trip(key, user_id, trip_id):
     # we know that the trip is cleaned so we can use the fmt_time
     # but the confirm objects are not necessarily filled out
     fmt_ts = lambda ts, tz: arrow.get(ts).to(tz)
+
+    if hasattr(tl_entry.data, 'start_ts'):
+        entry_start = tl_entry.data.start_ts
+    else:
+        entry_start = tl_entry.data.enter_ts
+    if hasattr(tl_entry.data, 'end_ts'):
+        entry_end = tl_entry.data.end_ts
+    else:
+        entry_end = tl_entry.data.exit_ts
+        # if a place has no exit time, the user hasn't left there yet
+        # so we will set the end time as high as possible for the purpose of comparison
+        entry_end = EPOCH_MAXIMUM if entry_end is None else entry_end
+
     logging.debug("Comparing user input %s: %s -> %s, trip %s -> %s, start checks are (%s && %s) and end checks are (%s || %s)" % (
         user_input.data.label,
         fmt_ts(user_input.data.start_ts, user_input.metadata.time_zone),
         fmt_ts(user_input.data.end_ts, user_input.metadata.time_zone),
-        trip_obj.data.start_fmt_time, trip_obj.data.end_fmt_time,
-        (user_input.data.start_ts >= trip_obj.data.start_ts),
-        (user_input.data.start_ts <= trip_obj.data.end_ts),
-        (user_input.data.end_ts <= trip_obj.data.end_ts),
-        ((user_input.data.end_ts - trip_obj.data.end_ts) <= 15 * 60)
+        fmt_ts(entry_start, user_input.metadata.time_zone), fmt_ts(entry_end, user_input.metadata.time_zone),
+        (user_input.data.start_ts >= entry_start),
+        (user_input.data.start_ts < entry_end),
+        (user_input.data.end_ts <= entry_end),
+        ((user_input.data.end_ts - entry_end) <= 15 * 60)
     ))
-    start_checks = (user_input.data.start_ts >= trip_obj.data.start_ts and
-        user_input.data.start_ts >= trip_obj.data.start_ts)
-    end_checks = (user_input.data.end_ts <= trip_obj.data.end_ts or
-        ((user_input.data.end_ts - trip_obj.data.end_ts) <= 15 * 60))
+    start_checks = (user_input.data.start_ts >= entry_start and
+        user_input.data.start_ts < entry_end)
+    end_checks = (user_input.data.end_ts <= entry_end or
+        ((user_input.data.end_ts - entry_end) <= 15 * 60))
     if start_checks and not end_checks:
         logging.debug("Handling corner case where start check matches, but end check does not")
-        next_trip_obj = _get_next_cleaned_trip(ts, trip_obj)
-        if next_trip_obj is not None:
-            end_checks = user_input.data.end_ts <= next_trip_obj.data.start_ts
-            logging.debug("Second level of end checks when the next trip is defined (%s <= %s) = %s" % (
-                user_input.data.end_ts, next_trip_obj.data.start_ts, end_checks))
+        next_entry_obj = _get_next_cleaned_timeline_entry(ts, tl_entry)
+        if next_entry_obj is not None:
+            if hasattr(next_entry_obj.data, 'end_ts'):
+                next_entry_end = next_entry_obj.data.end_ts
+            elif hasattr(next_entry_obj.data, 'exit_ts'):
```

Comment:
isn't this repeated from earlier? Refactor?

Reply:
I'll refactor into helper functions 'begin' and 'stop'